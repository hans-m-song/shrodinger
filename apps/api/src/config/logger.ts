import { Injectable, LoggerService, Scope } from '@nestjs/common';
import { LogFormat } from './init';
import pino from 'pino';
import { CoreError } from '@shrodinger/core/errors';
import init from '../config/init';

@Injectable({ scope: Scope.TRANSIENT })
export class Logger implements LoggerService {
  instance: pino.Logger;

  static create(context: string) {
    return new Logger(context);
  }

  constructor(readonly context: string) {
    this.instance = pino({
      level: init.log.level,
      name: context,
      formatters: {
        level: (label) => ({ level: label }),
      },
      serializers: {
        error: (error) => {
          if (typeof error !== 'object' || error === null) {
            return error;
          }

          if (CoreError.is(error)) {
            return error.toJSON();
          }

          const { name, message, stack } = error;
          return { name, message, stack };
        },
      },
      transport:
        init.log.format === LogFormat.Json
          ? undefined
          : {
              target: 'pino-pretty',
              options: {
                colorize: true,
              },
            },
    });
  }

  verbose(message: any, ...optionalParams: any[]) {
    this.instance.trace({}, message, ...optionalParams);
  }

  debug(message: any, ...optionalParams: any[]) {
    this.instance.debug({}, message, ...optionalParams);
  }

  log(message: any, ...optionalParams: any[]) {
    this.instance.info({}, message, ...optionalParams);
  }

  warn(message: any, ...optionalParams: any[]) {
    this.instance.warn({}, message, ...optionalParams);
  }

  error(message: any, ...optionalParams: any[]) {
    this.instance.error({}, message, ...optionalParams);
  }

  fatal(message: any, ...optionalParams: any[]) {
    this.instance.fatal({}, message, ...optionalParams);
  }
}
