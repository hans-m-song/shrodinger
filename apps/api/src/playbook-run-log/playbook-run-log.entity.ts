import { Field, Int, ObjectType } from '@nestjs/graphql';
import { AnsibleLogBase, PlaybookRunLog } from '@shrodinger/contracts';
import { GraphQLJSON } from 'graphql-scalars';
import { PlaybookRunLogAttributes } from '../database';

@ObjectType('PlaybookRunLog')
export class PlaybookRunLogEntity
  implements PlaybookRunLog, PlaybookRunLogAttributes
{
  @Field(() => Int)
  declare playbookRunId: number;

  @Field(() => Int)
  declare sequence: number;

  @Field(() => GraphQLJSON)
  declare contents: AnsibleLogBase & Record<string, unknown>;
}
