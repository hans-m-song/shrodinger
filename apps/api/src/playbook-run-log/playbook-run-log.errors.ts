import { createErrorSet } from '@shrodinger/core/errors';

export const PlaybookRunLogErrors = createErrorSet('PlaybookRunLog');
