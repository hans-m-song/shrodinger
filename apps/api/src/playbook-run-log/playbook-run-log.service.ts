import { Injectable } from '@nestjs/common';
import {
  ListPlaybookRunLogsHandler,
  ListPlaybookRunLogsQuery,
} from './handlers/list-playbook-run-logs.handler';
import {
  CreatePlaybookRunLogCommand,
  CreatePlaybookRunLogHandler,
} from './handlers/create-playbook-run-log.handler';

@Injectable()
export class PlaybookRunLogService {
  constructor(
    private readonly listPlaybookRunLogsHandler: ListPlaybookRunLogsHandler,
    private readonly createPlaybookRunLogHandler: CreatePlaybookRunLogHandler,
  ) {}

  async listPlaybookRunLogs(
    args: ConstructorParameters<typeof ListPlaybookRunLogsQuery>[0],
  ) {
    const query = new ListPlaybookRunLogsQuery(args);
    return this.listPlaybookRunLogsHandler.execute(query);
  }

  async createPlaybookRunLog(
    args: ConstructorParameters<typeof CreatePlaybookRunLogCommand>[0],
  ) {
    const command = new CreatePlaybookRunLogCommand(args);
    return this.createPlaybookRunLogHandler.execute(command);
  }
}
