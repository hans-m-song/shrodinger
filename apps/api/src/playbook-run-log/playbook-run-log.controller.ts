import { Controller } from '@nestjs/common';
import { InjectLogger, Logger } from '../logger';
import { PlaybookRunLogService } from './playbook-run-log.service';
import { tsRestHandler, TsRestHandler } from '@ts-rest/nest';
import { defaultResponse, playbookRunLogContract } from '@shrodinger/contracts';
import { Errors, HttpError } from '../../../../libs/core/errors';
import { ServerInferResponses } from '@ts-rest/core';

@Controller()
export class PlaybookRunLogController {
  constructor(
    @InjectLogger(PlaybookRunLogController.name)
    private readonly logger: Logger,
    private readonly playbookRunLogService: PlaybookRunLogService,
  ) {}

  @TsRestHandler(playbookRunLogContract.listPlaybookRunLogs)
  async listPlaybookRunLogs() {
    return tsRestHandler(
      playbookRunLogContract.listPlaybookRunLogs,
      async ({ params, query }) => {
        const result = await this.playbookRunLogService.listPlaybookRunLogs({
          ...params,
          ...query,
        });
        if (!result.ok) {
          this.logger.error(result.error);

          const error = Errors.as(HttpError)(result.error);
          if (error) {
            return error.toResponse() as ServerInferResponses<
              typeof playbookRunLogContract.listPlaybookRunLogs
            >;
          }

          return defaultResponse[500];
        }

        return {
          status: 200,
          body: { data: result.data },
        };
      },
    );
  }
}
