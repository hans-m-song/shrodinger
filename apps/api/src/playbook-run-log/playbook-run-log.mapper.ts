import { Injectable } from '@nestjs/common';
import {
  AnsibleLogEvent,
  AnsibleLogSchema,
  AnsiblePlaybookStats,
  AnsibleRunnerOkLogSchema,
  AnsibleRunnerLog,
  PlaybookRunLog,
} from '@shrodinger/contracts';
import { InjectLogger, Logger } from '../logger';
import { PlaybookRunLogErrors } from './playbook-run-log.errors';

export interface PlaybookRunLogTaskHost {
  hostname: string;
  duration: number;
  changed: boolean;
  data: Record<string, unknown>;
  deprecations?: string[];
  warnings?: string[];
  skip?: {
    reason: string;
    condition: string;
  };
}

export interface PlaybookRunLogTask {
  id: string;
  name: string;
  action: string;
  location: string;
  startedAt: number;
  hosts: Record<string, PlaybookRunLogTaskHost>;
}

export interface PlaybookRunResult {
  playbookName?: string;
  stats?: Record<string, AnsiblePlaybookStats>;
  tasks: PlaybookRunLogTask[];
}

@Injectable()
export class PlaybookRunLogMapper {
  constructor(
    @InjectLogger(PlaybookRunLogMapper.name)
    private readonly logger: Logger,
  ) {}

  public flatten(logs: PlaybookRunLog[]): PlaybookRunResult {
    const result: PlaybookRunResult = { tasks: [] };

    const tasks: Record<string, PlaybookRunLogTask> = {};

    for (const log of logs) {
      const parsed = AnsibleLogSchema.safeParse(log.contents);
      if (!parsed.success) {
        const error = new PlaybookRunLogErrors.Validation(
          parsed.error,
          log.contents,
        );

        this.logger.warn('could not parse ansible log line', error);
        continue;
      }

      switch (parsed.data._event) {
        case AnsibleLogEvent.PlaybookPlayStart:
          result.playbookName = parsed.data.play.name;
          continue;

        case AnsibleLogEvent.PlaybookStats:
          result.stats = parsed.data.stats;
          continue;

        case AnsibleLogEvent.PlaybookTaskStart:
          tasks[parsed.data.task.id] = {
            ...tasks[parsed.data.task.id],
            id: parsed.data.task.id,
            name: parsed.data.task.name,
            startedAt: parsed.data.task.duration.start,
          };
          continue;

        case AnsibleLogEvent.RunnerOk:
        case AnsibleLogEvent.RunnerSkipped:
          const host = this.flattenPlaybookRunRunnerLog(parsed.data);
          if (!host) {
            continue;
          }

          tasks[parsed.data.task.id] = {
            ...tasks[parsed.data.task.id],
            action: parsed.data.hosts[host.hostname].action,
            location: parsed.data.task.path,
            hosts: {
              ...tasks[parsed.data.task.id].hosts,
              [host.hostname]: host,
            },
          };
          continue;

        default:
          this.logger.warn('unknown event', { log });
          continue;
      }
    }

    result.tasks = Object.values(tasks).sort(
      (a, b) => a.startedAt - b.startedAt,
    );

    return result;
  }

  private flattenPlaybookRunRunnerLog(
    log: AnsibleRunnerLog,
  ): PlaybookRunLogTaskHost | null {
    const hostname = Object.keys(log.hosts)[0];
    const host = log.hosts[hostname];
    if (!hostname) {
      this.logger.warn('runner log did not contain any hosts', { log });
      return null;
    }

    AnsibleRunnerOkLogSchema;

    const {
      _ansible_no_log: _,
      _ansible_verbose_override: __,
      action: ___,
      skipped: ____,
      changed,
      deprecations,
      warnings,
      false_condition,
      skip_reason,
      ...data
    } = host as any;

    return {
      hostname,
      duration: log.task.duration.end - log.task.duration.start,
      changed,
      data,
      deprecations,
      warnings,
      skip: {
        condition: false_condition,
        reason: skip_reason,
      },
    };
  }
}
