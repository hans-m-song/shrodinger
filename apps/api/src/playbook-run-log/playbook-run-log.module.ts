import { forwardRef, Module } from '@nestjs/common';
import { DatabaseModule } from '../database';
import { PlaybookRunModule } from '../playbook-run/playbook-run.module';
import { PlaybookRunLogController } from './playbook-run-log.controller';
import { PlaybookRunLogService } from './playbook-run-log.service';
import { PlaybookRunLogResolver } from './playbook-run-log.resolver';
import { ListPlaybookRunLogsHandler } from './handlers/list-playbook-run-logs.handler';

@Module({
  imports: [DatabaseModule.register(), forwardRef(() => PlaybookRunModule)],
  controllers: [PlaybookRunLogController],
  providers: [
    PlaybookRunLogResolver,
    PlaybookRunLogService,
    ListPlaybookRunLogsHandler,
  ],
  exports: [PlaybookRunLogService],
})
export class PlaybookRunLogModule {}
