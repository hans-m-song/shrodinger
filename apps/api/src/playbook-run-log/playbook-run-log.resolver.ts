import { PlaybookRun, PlaybookRunLog } from '@shrodinger/contracts';
import { InjectLogger, Logger } from '../logger';
import { PlaybookRunService } from '../playbook-run/playbook-run.service';
import { PlaybookRunLogEntity } from './playbook-run-log.entity';
import { Args, Parent, Query, ResolveField, Resolver } from '@nestjs/graphql';
import { PlaybookRunLogService } from './playbook-run-log.service';
import { PlaybookRunEntity } from '../playbook-run/playbook-run.entity';
import { GraphQLError } from 'graphql';
import { ListPlaybookRunLogsArgs } from './handlers/list-playbook-run-logs.handler';

@Resolver(() => PlaybookRunLogEntity)
export class PlaybookRunLogResolver {
  constructor(
    @InjectLogger(PlaybookRunLogResolver.name)
    private readonly logger: Logger,
    private readonly playbookRunService: PlaybookRunService,
    private readonly playbookRunLogService: PlaybookRunLogService,
  ) {}

  @Query(() => [PlaybookRunLogEntity])
  async playbookRunLogs(
    @Args()
    args: ListPlaybookRunLogsArgs,
  ): Promise<PlaybookRunLog[]> {
    const result = await this.playbookRunLogService.listPlaybookRunLogs(args);
    if (!result.ok) {
      this.logger.error(result.error);
      throw new GraphQLError(result.error.message, {
        extensions: { code: result.error.name },
      });
    }

    return result.data;
  }

  @ResolveField(() => PlaybookRunEntity)
  async playbookRun(
    @Parent()
    { playbookRunId }: PlaybookRunLogEntity,
  ): Promise<PlaybookRun> {
    const result = await this.playbookRunService.getPlaybookRun({
      playbookRunId,
    });
    if (!result.ok) {
      this.logger.error(result.error);
      throw new GraphQLError(result.error.message, {
        extensions: { code: result.error.name },
      });
    }

    return result.data;
  }
}
