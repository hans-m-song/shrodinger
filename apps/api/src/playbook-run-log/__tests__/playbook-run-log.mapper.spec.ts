import { Logger } from '../../logger';
import { PlaybookRunLogMapper } from '../playbook-run-log.mapper';
import * as playbookRunLogFixture from '../../../test/fixtures/playbook-run-logs.json';
import { setLogging } from '../../../test/mocks/logger';

describe('PlaybookRunLogMapper', () => {
  const mapper = new PlaybookRunLogMapper(
    new Logger(PlaybookRunLogMapper.name),
  );

  describe('flatten', () => {
    it('should flatten playbook run logs', () => {
      setLogging(true);
      // given
      const logs = playbookRunLogFixture.map((log) => ({
        playbookRunId: 1,
        ...log,
      }));

      // when
      const result = mapper.flatten(logs);

      // then
      expect(result).toMatchSnapshot();
    });
  });
});
