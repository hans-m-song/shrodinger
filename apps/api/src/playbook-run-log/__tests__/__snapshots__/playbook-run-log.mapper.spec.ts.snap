// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaybookRunLogMapper flatten should flatten playbook run logs 1`] = `
{
  "playbookName": "Base",
  "stats": {
    "glados.local": {
      "changed": 0,
      "failures": 0,
      "ignored": 0,
      "ok": 9,
      "rescued": 0,
      "skipped": 1,
      "unreachable": 0,
    },
    "gman.local": {
      "changed": 0,
      "failures": 0,
      "ignored": 0,
      "ok": 9,
      "rescued": 0,
      "skipped": 1,
      "unreachable": 0,
    },
    "wheatley.local": {
      "changed": 0,
      "failures": 0,
      "ignored": 0,
      "ok": 9,
      "rescued": 0,
      "skipped": 1,
      "unreachable": 0,
    },
  },
  "tasks": [
    {
      "action": "gather_facts",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": [],
          "duration": 5645,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": [
            "Platform linux on host glados.local is using the discovered Python interpreter at /usr/bin/python3.10, but future installation of another Python interpreter could change the meaning of that path. See https://docs.ansible.com/ansible-core/2.17/reference_appendices/interpreter_discovery.html for more information.",
          ],
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": [],
          "duration": 4429,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": [
            "Platform linux on host gman.local is using the discovered Python interpreter at /usr/bin/python3.11, but future installation of another Python interpreter could change the meaning of that path. See https://docs.ansible.com/ansible-core/2.17/reference_appendices/interpreter_discovery.html for more information.",
          ],
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": [],
          "duration": 3904,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": [
            "Platform linux on host wheatley.local is using the discovered Python interpreter at /usr/bin/python3.11, but future installation of another Python interpreter could change the meaning of that path. See https://docs.ansible.com/ansible-core/2.17/reference_appendices/interpreter_discovery.html for more information.",
          ],
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-000000000029",
      "location": "/iac/ansible/playbooks/base/main.yaml:2",
      "name": "Gathering Facts",
      "startedAt": 1730538166497,
    },
    {
      "action": "ansible.builtin.debug",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 28,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 18,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 21,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-000000000006",
      "location": "/iac/ansible/playbooks/base/main.yaml:10",
      "name": "Debug",
      "startedAt": 1730538172168,
    },
    {
      "action": "ansible.builtin.package",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 40598,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 35679,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 23614,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-000000000007",
      "location": "/iac/ansible/playbooks/base/main.yaml:14",
      "name": "Install base packages",
      "startedAt": 1730538172201,
    },
    {
      "action": "ansible.builtin.apt_key",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 3469,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 3828,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 3250,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-00000000000e",
      "location": "/iac/ansible/roles/zerotier-client/tasks/main.yaml:2",
      "name": "zerotier-client : Add ZeroTier GPG apt key",
      "startedAt": 1730538212809,
    },
    {
      "action": "ansible.builtin.apt_repository",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 3157,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 2255,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 2161,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-00000000000f",
      "location": "/iac/ansible/roles/zerotier-client/tasks/main.yaml:7",
      "name": "zerotier-client : Add ZeroTier repository",
      "startedAt": 1730538216648,
    },
    {
      "action": "ansible.builtin.apt",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 3635,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 3555,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 2109,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-000000000010",
      "location": "/iac/ansible/roles/zerotier-client/tasks/main.yaml:17",
      "name": "zerotier-client : Install zerotier-one",
      "startedAt": 1730538219818,
    },
    {
      "action": "ansible.builtin.systemd",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 3299,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 2796,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 1903,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-000000000011",
      "location": "/iac/ansible/roles/zerotier-client/tasks/main.yaml:22",
      "name": "zerotier-client : Start zerotier-one",
      "startedAt": 1730538223457,
    },
    {
      "action": "ansible.builtin.command",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 26,
          "hostname": "glados.local",
          "skip": {
            "condition": "network_id is defined and network_id != """,
            "reason": "Conditional result was False",
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 22,
          "hostname": "gman.local",
          "skip": {
            "condition": "network_id is defined and network_id != """,
            "reason": "Conditional result was False",
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 22,
          "hostname": "wheatley.local",
          "skip": {
            "condition": "network_id is defined and network_id != """,
            "reason": "Conditional result was False",
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-000000000012",
      "location": "/iac/ansible/roles/zerotier-client/tasks/main.yaml:28",
      "name": "zerotier-client : Join ZeroTier network",
      "startedAt": 1730538226767,
    },
    {
      "action": "ansible.builtin.slurp",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 3538,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 2223,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 1699,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-000000000013",
      "location": "/iac/ansible/roles/zerotier-client/tasks/main.yaml:34",
      "name": "zerotier-client : Read ZeroTier client identity",
      "startedAt": 1730538226797,
    },
    {
      "action": "ansible.builtin.debug",
      "hosts": {
        "glados.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 201,
          "hostname": "glados.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "gman.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 201,
          "hostname": "gman.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
        "wheatley.local": {
          "changed": false,
          "data": {},
          "deprecations": undefined,
          "duration": 201,
          "hostname": "wheatley.local",
          "skip": {
            "condition": undefined,
            "reason": undefined,
          },
          "warnings": undefined,
        },
      },
      "id": "0691f17f-5892-5ccd-cc9d-000000000014",
      "location": "/iac/ansible/roles/zerotier-client/tasks/main.yaml:39",
      "name": "zerotier-client : Output ZeroTier client identity",
      "startedAt": 1730538230343,
    },
  ],
}
`;
